### PR TITLE
solaris_zone: fix zone configuration with python3

### LINKED ISSUE #1081: solaris_zone: zone configuration fails with python3
##### SUMMARY

Type error when trying to create a Solaris Zone with the `solaris_zone` module:

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

`solaris_zone`

##### ANSIBLE VERSION

```paste below
ansible 2.10.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/jbronn/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jbronn/.local/lib/python3.8/site-packages/ansible
  executable location = /home/jbronn/.local/bin/ansible-playbook
  python version = 3.8.5 (default, Jul 28 2020, 12:59:40) [GCC 9.3.0]
```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
OmniOS CE r151034t (Illumos); target host Python is 3.7.5.

##### STEPS TO REPRODUCE

```yaml
- name: 'solaris zones'
  hosts: all
  become: true
  tasks:
    - solaris_zone:
        name: z1
        state: installed
        path: /zones/z1
  vars:
    ansible_python_interpreter: '/usr/bin/python3'
```

##### EXPECTED RESULTS

The zone, `z1`, should be configured and installed.

##### ACTUAL RESULTS
Running the playbook produces this traceback:

```
Traceback (most recent call last):
  File "/export/home/jbronn/.ansible/tmp/ansible-tmp-1602198687.9610054-1444903-128778670541170/AnsiballZ_solaris_zone.py", line 102, in <module>
    _ansiballz_main()
  File "/export/home/jbronn/.ansible/tmp/ansible-tmp-1602198687.9610054-1444903-128778670541170/AnsiballZ_solaris_zone.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/export/home/jbronn/.ansible/tmp/ansible-tmp-1602198687.9610054-1444903-128778670541170/AnsiballZ_solaris_zone.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible_collections.community.general.plugins.modules.solaris_zone', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib/python3.7/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_solaris_zone_payload_p87jcqod/ansible_solaris_zone_payload.zip/ansible_collections/community/general/plugins/modules/solaris_zone.py", line 486, in <module>
  File "/tmp/ansible_solaris_zone_payload_p87jcqod/ansible_solaris_zone_payload.zip/ansible_collections/community/general/plugins/modules/solaris_zone.py", line 468, in main
  File "/tmp/ansible_solaris_zone_payload_p87jcqod/ansible_solaris_zone_payload.zip/ansible_collections/community/general/plugins/modules/solaris_zone.py", line 395, in state_present
  File "/tmp/ansible_solaris_zone_payload_p87jcqod/ansible_solaris_zone_payload.zip/ansible_collections/community/general/plugins/modules/solaris_zone.py", line 202, in configure
  File "/usr/lib/python3.7/tempfile.py", line 481, in func_wrapper
    return func(*args, **kwargs)
TypeError: a bytes-like object is required, not 'str'
```

### PR DESCRIPTION
**Pull request by jbronn**
##### SUMMARY

Fixes #1081 by explicitly opening `tempfile.NamedTemporaryFile` in text mode and adds a test.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

`community.general.solaris_zone`

##### ADDITIONAL INFORMATION

N/A

### CODE REVIEW

File: changelogs/fragments/1081-solaris_zone-python3.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+---
+bugfixes:
+  - solaris_zone - fixed issue trying to configure zone in python3 (https://github.com/ansible-collections/community.general/issues/1081)
```
Comment (felixfontein) [inline changelogs/fragments/1081-solaris_zone-python3.yml:3]:
```suggestion
  - solaris_zone - fixed issue trying to configure zone in Python 3 (https://github.com/ansible-collections/community.general/issues/1081).
```

File: plugins/modules/system/solaris_zone.py [hunk 2]
```diff
@@ -193,7 +193,7 @@ def configure(self):
             self.module.fail_json(msg='Missing required argument: path')
 
         if not self.module.check_mode:
-            t = tempfile.NamedTemporaryFile(delete=False)
+            t = tempfile.NamedTemporaryFile(delete=False, mode='wt')
```
Comment (felixfontein) [inline plugins/modules/system/solaris_zone.py:196]: Why not:
```suggestion
            t = tempfile.NamedTemporaryFile(delete=False, mode='w+t')
```
? That would be closer to the default value of `w+b` (https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile).

Comment (pmarkham) [reply plugins/modules/system/solaris_zone.py:196]: The "+" indicates, it's opened for reading and writing, which is not the case here. It only needs to be opened for writing. which is the correct option IMO.

Comment (jbronn) [reply plugins/modules/system/solaris_zone.py:196]: I agree with @pmarkham, the module is creating a file that's consumed by `zoneadm -f` and thus you don't need to read on the handle; but I'll change to `w+t` if that's what's required for merge.

Comment (felixfontein) [reply plugins/modules/system/solaris_zone.py:196]: I'm happy with the explanation, I just wanted to ask whether this is intentional or not :) No need to change it.

**Comment by felixfontein** [comment]
@pmarkham the PR looks good to me, though I'm not using Solaris. Since you are a maintainer of this module, please tell if you think this should be merged.

**Comment by pmarkham** [comment]
@felixfontein I don't currently have ready access to a Solaris host so I can't run any tests myself; however, looking at the PR @jbronn raised, I say go ahead and merge it.

**Comment by jbronn** [comment]
FWIW I had used this patch to provision my zones prior to submission here on OmniOS r151034t; should be fine to test on Solaris 11.3/4 as long as `runtime/python-34` or `runtime/python-35` is installed and used for the ansible interpreter.

In addition, while unit test cases I submitted doesn't actually invoke `zoneadm`, it does cover temporary file branch at issue -- we get some Solaris testing coverage using Linux containers.

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@jbronn thanks for fixing this!
@pmarkham thanks for reviewing!

**Pull request merged**